### ISSUE TITLE
Bug: <input type="number"> resets value if @value or @defaultValue is provided

### ISSUE THREAD

**Issue by ixrock** [facebook/react#8717]
**Demo:**
http://codepen.io/anon/pen/LxpyLB

**Problem:**
`<input type=number>` resets its value when I type "." instead of default behaviour (trigger input::invalid html5 state).

**Source code if demo not available:**
```html
<!-- demo.html -->
<p><b>Bug:</b></p>
<div id="test"></div>

<p><b>Normal html5 behavior:</b></p>
<div>
  <input type="number" value="20" /><br/>
  <input type="number" />
</div>
```

```css
/* demo.css */
input:invalid {
  outline: 1px solid red;
}
```

```js
// demo.js
class Test extends React.Component {
  public state = {value: 11};
  
  render(){
    return (
      <div>
      <p>1: 
      <input type="number" defaultValue="10"/>
      <input type="number" value={this.state.value} onChange={e => this.setState({value: e.target.value})}/>
      fully resets value when type "."</p>
      <p>2: <input type="number"/> allows to type "." and hightlight invalid field</p>
      </div>
    )
  }
}
ReactDOM.render(<Test/>, document.querySelector('#test'))
```

**Comment by Andarist** [comment]
Dibs! Gonna play around with it and see what I can do to fix the issue.

**Comment by Andarist** [comment]
So my investigation shows that... it's already fixed with [this commit](https://github.com/facebook/react/commit/0d20dcf9108811f632bfeb76a6bd3bf05d11865b). Although unfortunately it is not released yet.

**Comment by ixrock** [comment]
@Andarist it seems like that commit might fix a problem only with uncontrolled input, but the issue happens as well with controlled inputs, when `@value and @onChange` are used. 
I've updated the demo and code in first post to show this.

**Comment by Andarist** [comment]
Yeah, you are right, but I think there is nothing to be fixed on `react`'s side here, as controlled inputs depends on setting controlled values directly on the DOM nodes.

Check out [this demo](https://jsfiddle.net/p75syu09/) and inspect the `console`. You'll see a warning (in Chrome) that:
```
The specified value "20." is not a valid number. The value must match to the following regular expression: -?(\d+|\d+\.\d+|\.\d+)([eE][-+]?\d+)?
```

It seems that `react` has already experienced similar issue [here](https://github.com/facebook/react/issues/7487) and it was reported (and fixed) in the Chromium's bugtracker [here](https://bugs.chromium.org/p/chromium/issues/detail?id=638455). I propose to report this one as well so it can be fixed.

**Comment by gaearon** [comment]
cc @nhunzaker who worked on related things before.

**Comment by nhunzaker** [comment]
Thanks!

> So my investigation shows that... it's already fixed with this commit. Although unfortunately it is not released yet.

Yes! @gaearon we could probably cherry pick this to 15.x pretty safely. What do you think?

> @Andarist it seems like that commit might fix a problem only with uncontrolled input, but the issue happens as well with controlled inputs, when @value and @onChange are used.

@ixrock Yes! I've been maintaining a PR for controlled inputs: https://github.com/facebook/react/pull/7359. 

>  I propose to report this one as well so it can be fixed.

@Andarist I think this is a fantastic next step. I've also encountered this behavior in Safari 10.x as well. We should report it to them too. Happy to help clarify anything to help issue submission.

**Comment by gaearon** [comment]
>Yes! @gaearon we could probably cherry pick this to 15.x pretty safely. What do you think?

This commit was in 15.4.2, no?

**Comment by gaearon** [comment]
Argh, I forgot to mention it in the changelog. Sorry.

**Comment by gaearon** [comment]
Updated the changelog. ed760d15676edac4a2f565fc02460b022001ffa4

**Comment by nhunzaker** [comment]
Anything else to do here?

**Comment by aweary** [comment]
Should be good to close, the behavior appears to be fixed in 15.4.2
